Windows Support

rafaelsamenezes opened the pull request:
# Summary

This intends to add initial build support for Windows. There are some caveats and issues that I will go through in this description. But essentially, this PR will make ESBMC compile on x64 Windows with MSVC and Z3.

# Notes

1. The models are not correctly converted if the *.hs* files are in CRLF.
2. Because of the interaction of having to precompile LLVM and Boost I found it simpler and easier to make it dynamic and ship the DLL's
3. Z3 is chosen because it has an official prebuilt that keeps things simple.
4. The old-frontend is disabled. I don't think that we should waste time making it compatible.
5. The current prebuild of LLVM used is built in Release mode. So the only supported build mode is Release unless we rebuild it with Debug info. 
6. Some tests take too long under Z3. They are marked as throughout now.
7. Some tests are broken at Z3 (even under Linux), they where moved into another regression (nonz3)

# How to check this PR.

Since this contains lots of changes I will make an order suggestion:

1. `.gitattributes` and `.gitignore` should be straightforward
1. The new module for OSConfiguration at `scripts/cmake`: AppleConfiguration, WindowsConfiguration, and UnixConfiguration
1. General CMake changes, the usage of `target_precompiled_headers` , installation of DLL's, and new Licenses
1. ESBMC's code fixes. Most of them are to wrap `unistd.h` and POSIX calls, some fixes were needed for MSVC e.g. adding `extern` in `globals.cpp`.
1. libm changes (WIP). 
1. Build Instructions for Windows - hosted at the Wiki page
1. CI/CD scripts

rafaelsamenezes committed:
Marked .hs and .sh files as LF

rafaelsamenezes committed:
Update .gitignore

rafaelsamenezes committed:
Fixed boost libraries issue

rafaelsamenezes committed:
Marked irep2.h as precompiled

rafaelsamenezes committed:
Marked irep2.h as precompiled

rafaelsamenezes committed:
Marked irep2.h as precompiled

rafaelsamenezes committed:
Marked irep2.h as precompiled

rafaelsamenezes committed:
Removed shell command

rafaelsamenezes committed:
Removed shell command

rafaelsamenezes committed:
Added unistd.h on ifdef

rafaelsamenezes committed:
Added OSConfiguration modules

rafaelsamenezes committed:
Added OSConfiguration modules

Commit:
CMake now uses buildidobj.py

Signed-off-by: Rafael Sá Menezes <rafael.sa@icomp.ufam.edu.br>

rafaelsamenezes committed:
windows fixes in models

rafaelsamenezes committed:
Added precompiled headers

rafaelsamenezes committed:
Added precompiled headers

rafaelsamenezes committed:
Added precompiled headers

rafaelsamenezes committed:
Added precompiled headers

rafaelsamenezes committed:
Windows-specific code fixes

rafaelsamenezes committed:
msvc fixes

rafaelsamenezes committed:
msvc fixes

rafaelsamenezes committed:
windows build will now install dll's

rafaelsamenezes committed:
Create BOOST_LICENSE.txt

rafaelsamenezes committed:
Added /W1 for msvc

rafaelsamenezes committed:
Added /W1 for msvc

rafaelsamenezes committed:
Added boost libraries into target

rafaelsamenezes committed:
Initial windows CI

rafaelsamenezes committed:
Initial windows CI

rafaelsamenezes committed:
Added strip to make CRLF/LF compatible

rafaelsamenezes committed:
Added strip to make CRLF/LF compatible

rafaelsamenezes committed:
Added hack for windows python

rafaelsamenezes committed:
Added hack for windows python

rafaelsamenezes committed:
Added Irep2Optmization script

rafaelsamenezes committed:
Added Irep2Optmization script

rafaelsamenezes committed:
Made release default windows build

rafaelsamenezes committed:
Added /bigobj in MSVC

rafaelsamenezes committed:
Marked extern on ifdef WIN32

rafaelsamenezes committed:
Fixed cmake command

rafaelsamenezes committed:
Added msbuild command

rafaelsamenezes committed:
Replaced msbuild with cmake equivalent

rafaelsamenezes committed:
Added OLD_FRONTEND_TARGETS

rafaelsamenezes committed:
Added windows artifact generation

rafaelsamenezes committed:
Fixed bad ident

rafaelsamenezes committed:
Enabled tests for windows

rafaelsamenezes committed:
Removed unused warning

rafaelsamenezes committed:
Added Z3 and Boost into windows

rafaelsamenezes committed:
Fixed ctest command

rafaelsamenezes committed:
Enabled regression on Windows CI

rafaelsamenezes committed:
Update build.yml

Commit:
Code-style fixes

Commit:
Marked tests as Linux only

Commit:
Added parallel test execution in Windows

rafaelsamenezes committed:
Marked all files as LF

mikhailramalho commented:
I saw that you removed `thread.cpp` from the compilation and, given that it is not used anywhere and that there are standard alternatives that work on Linux/Mac/Windows (i.e., `std::thread`), I'm OK with killing it altogether.

rafaelsamenezes committed:
Update linter.yml

rafaelsamenezes commented:
> I saw that you removed `thread.cpp` from the compilation and, given that it is not used anywhere and that there are standard alternatives that work on Linux/Mac/Windows (i.e., `std::thread`), I'm OK with killing it altogether.

I forgot about that. I agree

rafaelsamenezes committed:
Fix for CRLF endings

rafaelsamenezes committed:
Marked boolector tests as throrough

rafaelsamenezes committed:
Made windows-only hacks

rafaelsamenezes committed:
Added commands to unix

rafaelsamenezes committed:
Added commands to unix

rafaelsamenezes committed:
Fixed issue that c2goto input arguments

rafaelsamenezes committed:
Fixed issue that c2goto input arguments

Commit:
Enabled libm

Commit:
Enabled libm

rafaelsamenezes committed:
Libm fixes for Windows

rafaelsamenezes committed:
Libm fixes for Windows

Commit:
Code-style fixup

Commit:
Added check for cmake version

rafaelsamenezes committed:
Added windows defines on clang

rafaelsamenezes committed:
Added windows defines on clang

rafaelsamenezes committed:
Fixed infinity testcases

rafaelsamenezes committed:
Fixed infinity testcases

rafaelsamenezes committed:
marked new nonz3 test

rafaelsamenezes committed:
Updated float tests

rafaelsamenezes committed:
Fixes for windows

rafaelsamenezes committed:
Fixes for windows

rafaelsamenezes committed:
Fixed testcase

rafaelsamenezes committed:
Added __builtin_signbit

rafaelsamenezes committed:
Added __builtin_signbit

rafaelsamenezes committed:
Update configure_environment.ps1

rafaelsamenezes committed:
marked unix tests

rafaelsamenezes committed:
marked z3 test as thorough

rafaelsamenezes committed:
Added Windows regressions

rafaelsamenezes committed:
marked tests as nonz3

rafaelsamenezes committed:
changed windows regressions

rafaelsamenezes committed:
Update build.yml

rafaelsamenezes committed:
moved windows tests and nonz3 tests

Tests that don't work on windows because of the redefinition of size_t are marked as throrough and the tests that trigger issues on Z3 are marked as nonz3

rafaelsamenezes committed:
code-style fixes

rafaelsamenezes committed:
removed flail test

rafaelsamenezes committed:
removed clone as LF in ubuntu

rafaelsamenezes committed:
Update fpclassify.c

rafaelsamenezes committed:
added execution permission for testing tool

rafaelsamenezes committed:
boost is now copied from vcpkg

rafaelsamenezes committed:
fixed code-style

rafaelsamenezes committed:
moved configure_environment to before cmake

rafaelsamenezes committed:
uncommented typedef

rafaelsamenezes committed:
fixed boost version

rafaelsamenezes committed:
/s/Deps/deps

rafaelsamenezes committed:
updated windows build

rafaelsamenezes committed:
updated boost to version 1_75

rafaelsamenezes committed:
updated windows build

rafaelsamenezes committed:
moved testcases to nonz3

rafaelsamenezes committed:
removed thread files

mikhailramalho commented on regression/esbmc/no_pointer_check_1/main.c:
```diff
@@ -1,8 +1,8 @@
+
+    if(p < &q) return 1;
+    return 0;
+}
```
what are the changes to these files?

rafaelsamenezes replied:
I think they got in the middle of a change when I was forcing everything to be LF. So it is changing from CRLF to LF. I am not sure the proper way to handle this. Do we let it, or should I revert this? If needed, could you help me reverting it?

mikhailramalho commented on scripts/configure_environment.ps1:
```diff
@@ -0,0 +1,17 @@
+# Download and extract prebuilt LLVM+Clang for Windows
+
+$URL = "https://www.dropbox.com/s/z1gyschfa46yj6e/clang.zip?dl=1"
```
can we use the official clang download link instead of a dropbox link?

rafaelsamenezes replied:
No, the official clang for Windows does not ship with all the needed libraries (and neither the CMake configuration)

mikhailramalho commented on src/c2goto/library/libm/fpclassify.c:
```diff
@@ -1,9 +1,35 @@
+  int name(type d)                                                             \
+  {                                                                            \
+  __ESBMC_HIDE:;                                                               \
+    \ 
```
extra backslash?

rafaelsamenezes replied:
Removed it

mikhailramalho commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -423,6 +426,9 @@ int esbmc_parseoptionst::doit()
 {
+#ifdef _WIN32
+  std::cerr << "Windows does not support parallel kind" << std::endl;
+#else
```
I suggest removing the `#else` and aborting here.

Also, use '\n' instead of `std::endl`:

rafaelsamenezes replied:
> I suggest removing the #else and aborting here.

Ok

> Also, use '\n' instead of std::endl

Why? Windows uses CRLF

rafaelsamenezes replied:
> I suggest removing the #else and aborting here.

@mikhailramalho  actually I can't do this, the `else` contains unix system calls i.e `pipe`

mikhailramalho commented on src/esbmc/globals.cpp:
```diff
@@ -10,4 +10,8 @@ const mode_table_et mode_table[] = {
+#ifdef _WIN32
+extern "C" uint8_t *esbmc_version_string = buildidstring_buf;
+#else
 uint8_t *esbmc_version_string = buildidstring_buf;
```
do we need this ifdef here? If the windows version work on Linux (with the extern C), we should keep only that.

rafaelsamenezes replied:
Yes, GCC will complain about this and I didn't find a good way to disable the warning for external initialization

mikhailramalho commented on src/pointer-analysis/value_set.h:
```diff
@@ -17,6 +17,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
+#undef min
+#include <algorithm>
+
```
Why do we need this here? Can you at least move this to a c file? Since it is in a header, it may impact several other translation units.

mikhailramalho commented on src/solvers/smtlib/smtlib_conv.cpp:
```diff
@@ -160,7 +161,10 @@ smtlib_convt::smtlib_convt(bool int_encoding, const namespacet &_ns)
+#ifdef _WIN32
+  // TODO: The current implementation uses UNIX Process
+  std::cerr << "smtlib works only in unix systems" << std::endl;
+#else
```
Remove #else and abort.

rafaelsamenezes replied:
> Remove #else 

Same as above, the else uses unix syscalls

> ...and abort

Ok

mikhailramalho commented on src/solvers/smtlib/smtlib_conv.h:
```diff
@@ -4,7 +4,11 @@
 #include <string>
+#ifdef _WIN32
+typedef uint64_t pid_t;
+#else
```
I think we can remove this and delete the pid_t member from this class.

rafaelsamenezes replied:
I think the member is used for unix.

mikhailramalho reviewed (changes requested):
Most of it looks good, however, I have a few questions/concerns:
1. Why some many files where changed but look the same (maybe the github UI is not showing all the changes).
2. what's the deal with the undef min?

mikhailramalho commented:
Because endl flushes the output, and that's probably something you didn't
intend to do here.

Windows terminal is clever enough to understand only the \n.

Em sáb., 30 de jan. de 2021 às 12:48, Rafael Sá Menezes <
notifications@github.com> escreveu:

> *@rafaelsamenezes* commented on this pull request.
> ------------------------------
>
> In src/esbmc/esbmc_parseoptions.cpp
> <https://github.com/esbmc/esbmc/pull/282#discussion_r567274897>:
>
> > @@ -423,6 +426,9 @@ int esbmc_parseoptionst::doit()
>
>  int esbmc_parseoptionst::doit_k_induction_parallel()
>  {
> +#ifdef _WIN32
> +  std::cerr << "Windows does not support parallel kind" << std::endl;
> +#else
>
> I suggest removing the #else and aborting here.
>
> Ok
>
> Also, use '\n' instead of std::endl
>
> Why? Windows uses CRLF
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/282#discussion_r567274897>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH6EEGHXIQ3OVUZRMDTS4QZ7DANCNFSM4SAMT25Q>
> .
>


-- 

Mikhail Ramalho.

rafaelsamenezes committed:
removed extra backslash

rafaelsamenezes committed:
removed undef min

rafaelsamenezes committed:
replace endl to \n and added abort

rafaelsamenezes committed:
code-style fixes

rafaelsamenezes committed:
removed pid_t member from smtlib_conv

rafaelsamenezes committed:
marked testcase as CRLF

mikhailramalho commented on regression/CMakeLists.txt:
```diff
@@ -14,17 +14,20 @@ function(add_esbmc_regression folder mode)
+    set(REGRESSIONS esbmc cbmc cstd llvm floats floats-regression k-induction esbmc-unix esbmc-unix2 k-induction-parallel nonz3)
+elseif(WIN32)
+    # FUTURE: Add floats-regression esbmc-cpp/cpp
+    set(REGRESSIONS esbmc cbmc cstd llvm floats  k-induction )
```
Why isn't the floats-regression included here?

rafaelsamenezes replied:
because the models for it are not finished yet. I intend to open an issue for this

mikhailramalho replied:
I see, like the signbit one? I agree, they should come in a separate PR.

mikhailramalho commented on regression/nonz3/matrix_bug/main.c:
```diff
@@ -1,35 +1,35 @@
+//  Aluno: Fredson Souza da Costa  Matr�cula: 20902628
+//  Verifica��o de propriedades de seguran�a
+//  programa: I) ENCONTRA O MAIOR ELEMENTO EM UMA MATRIZ
+
```
encoding is broken

mikhailramalho commented on scripts/cmake/WindowsConfiguration.cmake:
```diff
@@ -0,0 +1,36 @@
+  if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+	  # There are a LOT of warnings from clang headers
+	  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-everything")
+	  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-everything")
```
Can we be more precise about what warning we are disabling here?

rafaelsamenezes replied:
Since my focus was with MSVC I didn't looked very much at the clang compiler. However, it was from various domains and it stopped the compilation process

mikhailramalho replied:
Right, I get it now. Does our regression build ESBMC with MSVC?

rafaelsamenezes replied:
Yes it does

mikhailramalho reviewed (changes requested):
A couple of small changes required now.

rafaelsamenezes committed:
fixed enconding
